src/python/pants/backend/adhoc/code_quality_tool.py

```diff
@@ -0,0 +1,549 @@
+            )
+
+            proc_result = await Get(
+                FallibleProcessResult,
+                CodeQualityToolBatch(
+                    runner=code_quality_tool_runner,
+                    sources_snapshot=sources_snapshot,
+                    output_files=(),
+                ),
+            )
+
+            return LintResult.create(request, process_result=proc_result)
+
+        namespace = dict(locals())
+
+        return [
+            *collect_rules(namespace),
+            *CodeQualityProcessingRequest.rules(),
+        ]
+
+    def _build_fmt_rules(self) -> Iterable[Rule]:
+        class CodeQualityToolInstance(Subsystem):
+            options_scope = self.scope
+            name = self.name
+            help = f"{self.goal.capitalize()} with {self.name}. Tool defined in {self.target}"
+
+            skip = SkipOption("lint", "fmt")
+
+        class CodeQualityProcessingRequest(FmtFilesRequest):
+            tool_subsystem = CodeQualityToolInstance
+
+        @rule(canonical_name_suffix=self.scope)
+        async def partition_inputs(
+            request: CodeQualityProcessingRequest.PartitionRequest,
+            subsystem: CodeQualityToolInstance,
+        ) -> Partitions:
+            if subsystem.skip:
+                return Partitions()
+
+            cqt = await Get(CodeQualityTool, CodeQualityToolAddressString(address=self.target))
+
+            matching_filepaths = FilespecMatcher(
+                includes=cqt.file_glob_include,
+                excludes=cqt.file_glob_exclude,
+            ).matches(request.files)
+
+            return Partitions.single_partition(sorted(matching_filepaths))
+
+        @rule(canonical_name_suffix=self.scope)
+        async def run_code_quality(request: CodeQualityProcessingRequest.Batch) -> FmtResult:
+            sources_snapshot = request.snapshot
+
+            code_quality_tool_runner = await Get(
+                CodeQualityToolBatchRunner, CodeQualityToolAddressString(address=self.target)
+            )
+
+            proc_result = await Get(
+                FallibleProcessResult,
+                CodeQualityToolBatch(
+                    runner=code_quality_tool_runner,
+                    sources_snapshot=sources_snapshot,
+                    output_files=request.files,
+                ),
+            )
+
+            output = await Get(Snapshot, Digest, proc_result.output_digest)
+
+            return FmtResult(
+                input=request.snapshot,
+                output=output,
+                stdout=Simplifier().simplify(proc_result.stdout),
+                stderr=Simplifier().simplify(proc_result.stderr),
+                tool_name=request.tool_name,
+            )
+
+        namespace = dict(locals())
+
+        return [
+            *collect_rules(namespace),
+            *CodeQualityProcessingRequest.rules(),
+        ]
+
+    def _build_fix_rules(self) -> Iterable[Rule]:
+        class CodeQualityToolInstance(Subsystem):
```

Comment:
Classes defined inside of functions can have goofy/unhelpful names, like `foo.<locals>.Bar`. You can override `__name__` and `__qualname__` to change that. I think we do that in a few other class-generation cases.

Reply:
Yes, they end up having those names. Let me see what I can dig up to improve it... it looks like we do it in `distinct_union_type_per_subclass` at least. Let me see if I can have it surface in a test and then will adjust it.

Reply:
Ah, made an attempt but a significant complication arises: a part of the options bootstrapping mechanism inspects the source code AST to find the class (for reasons I am not entirely sure of, and had to disable in previous iterations). If we dynamically change the qualname, that logic becomes unable to find the class.

https://github.com/pantsbuild/pants/blob/a6c55c149918de2893a973225b10b7b1c4d60294/src/python/pants/option/subsystem.py#L197-L198

Reply:
oh, interesting. I guess that's why `distinct_union_type_per_subclass` uses the 3-arg form of `type`